### DISCUSSION TITLE
MvcMovieGiraffe

### DISCUSSION THREAD

**Discussion by dharmatech** [giraffe-fsharp/Giraffe#456] [Show and tell]
The ASP.NET Core documentation includes an MVC tutorial for building a project called [MvcMovies](https://docs.microsoft.com/en-us/aspnet/core/tutorials/first-mvc-app/?view=aspnetcore-5.0).

[MvcMovieGiraffe](https://github.com/dharmatech/mvc-movie-giraffe) is a port of that project to F#/Giraffe.

It's very rough around the edges but it runs and the various pages are there. It currently uses the in-memory database.

**Comment by dustinmoris** [comment]
Awesome, thanks for bringing this to my attention and thank you for your support! Is this project something you would like to add to the current [collection of Giraffe samples](https://github.com/giraffe-fsharp/samples)? At least we should at a link to your repository in the README of the samples!

**Reply by dharmatech** [reply]
> At least we should at a link to your repository in the README of the samples!

Sure, feel free to mention it in the README if you'd like. :-)

**Comment by dharmatech** [comment]
# Refactoring views

The original C# version of MvcMovie has the following in the Edit view:

```html
<div class="form-group">
    <label asp-for="Title" class="control-label"></label>
    <input asp-for="Title" class="form-control" />
    <span asp-validation-for="Title" class="text-danger"></span>
</div>

<div class="form-group">
    <label asp-for="ReleaseDate" class="control-label"></label>
    <input asp-for="ReleaseDate" class="form-control" />
    <span asp-validation-for="ReleaseDate" class="text-danger"></span>
</div>

<div class="form-group">
    <label asp-for="Genre" class="control-label"></label>
    <input asp-for="Genre" class="form-control" />
    <span asp-validation-for="Genre" class="text-danger"></span>
</div>

<div class="form-group">
    <label asp-for="Price" class="control-label"></label>
    <input asp-for="Price" class="form-control" />
    <span asp-validation-for="Price" class="text-danger"></span>
</div>

<div class="form-group">
    <label asp-for="Rating" class="control-label"></label>
    <input asp-for="Rating" class="form-control" />
    <span asp-validation-for="Rating" class="text-danger"></span>
</div>
```

Now that the experimental tag helpers are in place, the [corresponding code](https://github.com/dharmatech/mvc-movie-giraffe/blob/e44937b4368d761c88bffe1a0dac63a90ea7ac90/Program.fs#L354-L368) in the Giraffe version is:

```fsharp
let form_group (expr : FSharp.Quotations.Expr<'a>) =
    div [ _class "form-group" ] 
        [
            TagHelpers.Label.Of(%expr, [ _class "control-label" ])

            TagHelpers.Input.Of(%expr, [ _class "form-control" ])

            TagHelpers.SpanValidation.Of(%expr, [ _class "text-danger" ])
        ]

form_group <@ model.Title @>
form_group <@ model.ReleaseDate @>
form_group <@ model.Genre @>
form_group <@ model.Price @>
form_group <@ model.Rating @>
```
